### PR TITLE
service: handle API error properly

### PR DESCRIPTION
**Pull request by BusyJay**
Errors returned by storage API calls should be handled properly,
instead of turning them into grpc errors, which will affect how does
the TiClient choose retry policy.

Note that ServerBusy is reported when the channel of scheduler is
full from now on. And when threaded raftstore is landed, there may
be more region related errors reported by the API calls.

/cc pingcap/tidb#7354

### CODE REVIEW

File: src/util/future.rs [hunk 1]
```diff
@@ -29,3 +31,62 @@ where
     };
     (callback, future)
 }
+
+/// A shortcut for `f1.and_then(|()| f2.flatten())`. Note that
+/// the expression is just a simplified version as f2's Error
+/// type may not be easy handled via combinators.
+pub struct AndThenWith<F1, F2>
```
Comment (siddontang) [inline src/util/future.rs]: where to use this?

Comment (BusyJay) [reply src/util/future.rs]: In `kv.rs`, where the result of API call needs to be handled the same way as the value returned by the future.

File: src/server/service/kv.rs [hunk 2]
```diff
@@ -859,31 +783,25 @@ impl<T: RaftStoreRouter + 'static, E: Engine> tikvpb_grpc::Tikv for Service<T, E
...
         );
-        if let Err(e) = res {
-            self.send_fail_status(ctx, sink, Error::from(e), RpcStatusCode::ResourceExhausted);
-            return;
-        }
 
-        let future = future
-            .map_err(Error::from)
-            .map(|v| {
+        let future = AndThenWith::new(res, f.map_err(Error::from))
```
Comment (breeswish) [inline src/server/service/kv.rs]: How about:

```rust
future::result(res)
    .map_err(Error::from)
    .and_then(|_| future.map_err(Error::from))
    .and_then(|...|)
```

like what we did in [storage](https://github.com/tikv/tikv/blob/master/src/storage/mod.rs#L547)?

Comment (BusyJay) [reply src/server/service/kv.rs]: That's wrong. `res` should be handled the same way as what return by the `future`. So this becomes `future::result(res).and_then(|_| future.flatten())` like what is [commented](https://github.com/tikv/tikv/pull/3450/files#diff-ce14a2c92d99351d9bbb883addee9878R35). All in all, using a custom future here is more expressive and clean.

Comment (BusyJay) [reply src/server/service/kv.rs]: It's approachable but I prefer this way as it's shorter.

**Comment by BusyJay** [comment]
PTAL

File: src/server/service/kv.rs [hunk 3]
```diff
@@ -915,6 +833,7 @@ impl<T: RaftStoreRouter + 'static, E: Engine> tikvpb_grpc::Tikv for Service<T, E
 
         let task = EndPointTask::Request(req_task);
         if let Err(e) = self.end_point_scheduler.schedule(task) {
+            // TODO: report busy instead.
```
Comment (breeswish) [inline src/server/service/kv.rs]: Why don't fix it by using this PR's approach?

Comment (BusyJay) [reply src/server/service/kv.rs]: They are not the same. Cases fixed above are raftstore errors, here is just the schedule error.

Comment (breeswish) [reply src/server/service/kv.rs]: Ok

**Pull request merged**